File: implementations/python/bfs/bfs_small.py
import sys
from collections import deque

N = 100

adj = [[False] * N for _ in range(N)]
visited = [False] * N

def setup():
    for i in range(N - 1):
        adj[i][i + 1] = True
        adj[i + 1][i] = True
    for i in range(0, N, 10):
        j = (i * 7 + 3) % N
        adj[i][j] = True
        adj[j][i] = True

def bfs():
    for i in range(N):
        visited[i] = False
    queue = deque()
    visited[0] = True
    queue.append(0)
    count = 0
    while queue:
        cur = queue.popleft()
        count += 1
        for nb in range(N):
            if adj[cur][nb] and not visited[nb]:
                visited[nb] = True
                queue.append(nb)
    return count

def main():
    setup()

    sys.stdout.write("ready\n")
    sys.stdout.flush()

    for line in sys.stdin:
        sys.stdout.write(str(bfs()) + "\n")
        sys.stdout.flush()

if __name__ == "__main__":
    main()

File: implementations/rust/bfs/small/bfs_small.rs
use std::io::{self, BufRead, Write};

const N: usize = 100;

static mut ADJ:     [[bool; N]; N] = [[false; N]; N];
static mut VISITED: [bool; N]      = [false; N];
static mut QUEUE:   [usize; N]     = [0; N];

fn setup() {
    unsafe {
        for i in 0..N - 1 {
            ADJ[i][i + 1] = true;
            ADJ[i + 1][i] = true;
        }
        for i in (0..N).step_by(10) {
            let j = (i * 7 + 3) % N;
            ADJ[i][j] = true;
            ADJ[j][i] = true;
        }
    }
}

fn bfs() -> usize {
    unsafe {
        for i in 0..N {
            VISITED[i] = false;
        }
        let mut head = 0;
        let mut tail = 0;
        let mut count = 0;
        VISITED[0]    = true;
        QUEUE[tail]   = 0;
        tail += 1;
        while head < tail {
            let cur = QUEUE[head];
            head += 1;
            count += 1;
            for nb in 0..N {
                if ADJ[cur][nb] && !VISITED[nb] {
                    VISITED[nb]  = true;
                    QUEUE[tail]  = nb;
                    tail += 1;
                }
            }
        }
        count
    }
}

fn main() {
    setup();

    println!("ready");
    io::stdout().flush().unwrap();

    let stdin = io::stdin();
    for _ in stdin.lock().lines() {
        println!("{}", bfs());
        io::stdout().flush().unwrap();
    }
}